device/common/include/traccc/clusterization/device/impl/reduce_problem_cell.ipp

```diff
@@ -70,14 +62,15 @@ inline void reduce_problem_cell(
          * Note that this check now looks in the opposite direction! An
          * important difference.
          */
-        if (cells[j].channel1 > c1 + 1 || cells[j].module_link != mod_id) {
+        if (traccc::details::is_far_enough(reference_cell, cells.at(j))) {
```

Comment:
I'll test it out tomorrow, but I'm pretty sure that this is the faulty line. 🤔 Cell order very much matters for this function call. And I believe the code currently uses the wrong one. Leading to the for-loop going on for much longer than it should.

Let's hope it will get most of the performance back at least...

Reply:
With this we are back to the numbers that I first reported. With the current code working like this:

![Screenshot from 2024-04-19 07-36-55](https://github.com/acts-project/traccc/assets/30694331/86b140bf-ae0e-4182-9566-543c01e489c1)

And this PR's code working like:

![Screenshot from 2024-04-19 07-39-12](https://github.com/acts-project/traccc/assets/30694331/16cc0708-5621-4188-b8fa-4767372807f9)

So `traccc::device::ccl_kernel(...)` is speeding up with this PR. But the seed finding is slowing down a little. Since the clusterization now reports to behave correctly, I can only imagine that the measurement/spacepoint order ending up differently would be the reason for this. But I'm not quite sure why. Why it would be such a large effect. Since it's not even the spacepoint binning that seems to slow down. :confused:

Reply:
Perfect, great work!